### PR TITLE
[Draft]: Add policy entropy metric

### PR DESCRIPTION
**Pull request by SamComber**
Would be good to have an option to understand potential policy entropy collapse. Exhaustion is typically indicative of growing determinism in token decoding and diminishes exploration behaviour (see https://arxiv.org/pdf/2505.22617 paper)

image

### WIP

Need to optimise so we don't create `B*L*V` intermediate tensors inside the entropy calc (i.e. basically 3x the memory required right now)... will likely batch it and put a warning flag when calculate entropy is enabled ??

### CODE REVIEW

File: verifiers/trainers/grpo_trainer.py [hunk 1]
```diff
@@ -731,18 +760,31 @@ def _get_per_token_logps(
...
             input_ids_batch = input_ids_batch[:, -logits_to_keep:]
             # For transformers<=4.48, logits_to_keep argument isn't supported, so here we drop logits ourselves.
             # See https://github.com/huggingface/trl/issues/2770
             logits = logits[:, -logits_to_keep:]
             # Divide logits by sampling temperature.
             # See https://huggingface.co/blog/the_n_implementation_details_of_rlhf_with_ppo#policy-training-implementation-details
             logits = logits / self.temperature
+
+            if compute_entropy:
+                entropy = entropy_from_logits_memory_efficient(logits, chunk_size=32)
```
Comment (qgallouedec) [inline verifiers/trainers/grpo_trainer.py]: hey! if you want something aligned with trl, we have `entropy_from_logits` that you can directly import

https://github.com/cyyever/trl/blob/2ddf8010881e3bdf215be28ed6fd6f3a8ae2bcf5/trl/trainer/utils.py#L1469

it does the same as here

Comment (SamComber) [reply verifiers/trainers/grpo_trainer.py]: Oh this is perfect! Thanks for the suggestion

Comment (SamComber) [reply verifiers/trainers/grpo_trainer.py]: Still getting OOMs with chunk size 1 here in places where we definitely should not be hmm @qgallouedec , does `entropy_from_logits` internally need to `torch.no_grad()` perhaps?
